File: Data Structures & Algorithms/combinations/submission-1.py
class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        res = []

        def rec(i, l):
            if len(l) == k:
                res.append(l[:])
                return
            if i > n or (n-i+1) < (k - len(l)):
                return
            l.append(i)
            rec(i+1, l)
            l.pop()
            rec(i+1, l)
        
        rec(1, [])
        return res

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-1.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Set<Character> windowSet = new HashSet<>();
        int res = 0;
        int l = 0;
        for (int r = 0; r < s.length(); r++) {
            while (windowSet.contains(s.charAt(r))) {
                windowSet.remove(s.charAt(l));
                l++;
            }
            windowSet.add(s.charAt(r));
            res = Math.max(res, r - l + 1);
        }
        return res;
    }
}
